src/Core/MySQL/Authentication.cpp

```diff
@@ -179,22 +169,36 @@ void Sha256Password::authenticate(
      */
     if (!is_secure_connection && !auth_response->empty() && auth_response != String("\0", 1))
     {
+        using EVP_PKEY_CTX_ptr = std::unique_ptr<EVP_PKEY_CTX, decltype(&EVP_PKEY_CTX_free)>;
+
         LOG_TRACE(log, "Received nonempty password.");
+
         const auto & unpack_auth_response = *auth_response;
         const auto * ciphertext = reinterpret_cast<const unsigned char *>(unpack_auth_response.data());
+        size_t ciphertext_len = unpack_auth_response.size();
 
-        std::vector<unsigned char> plaintext(RSA_size(&private_key));
-        int plaintext_size = RSA_private_decrypt(
-            static_cast<int>(unpack_auth_response.size()), ciphertext, plaintext.data(), &private_key, RSA_PKCS1_OAEP_PADDING);
-        if (plaintext_size == -1)
+        EVP_PKEY * pkey = static_cast<EVP_PKEY*>(private_key);
+        EVP_PKEY_CTX_ptr ctx(EVP_PKEY_CTX_new(pkey, nullptr), &EVP_PKEY_CTX_free);
+        if (!ctx)
+            throw Exception(ErrorCodes::OPENSSL_ERROR, "Failed to create EVP_PKEY_CTX");
+
+        if (EVP_PKEY_decrypt_init(ctx.get()) <= 0 ||
```

Comment:
For uniformity, we should use the same pattern as [here](https://github.com/ClickHouse/ClickHouse/pull/78808).

Reply:
```
EVP_PKEY_decrypt_init(), EVP_PKEY_decrypt_init_ex() and EVP_PKEY_decrypt() return 1 for success and 0 or a negative value for failure
```